rpc2/transport.go

```diff
@@ -20,215 +19,205 @@ type Transporter interface {
+	Encoder
+	sync.Locker
 }
 
-func (c *conPackage) Write(b []byte) (err error) {
-	_, err = c.con.Write(b)
-	return
-}
-
-func (c *conPackage) Close() error {
-	return c.con.Close()
-}
-
-func (c *conPackage) GetRemoteAddr() net.Addr {
-	return c.remoteAddr
+type connDecoder struct {
+	Decoder
+	net.Conn
+	io.ByteReader
 }
 
-type Transport struct {
-	mh         *codec.MsgpackHandle
-	cpkg       *conPackage
-	buf        *bytes.Buffer
-	enc        *codec.Encoder
-	mutex      *sync.Mutex
-	rdlck      *sync.Mutex
-	wrlck      *sync.Mutex
-	dispatcher dispatcher
-	packetizer *Packetizer
-	log        LogInterface
-	running    bool
-	wrapError  WrapErrorFunc
-}
-
-func NewconPackage(c net.Conn, mh *codec.MsgpackHandle) *conPackage {
+func newConnDecoder(c net.Conn, mh *codec.MsgpackHandle) *connDecoder {
 	br := bufio.NewReader(c)
 
-	return &conPackage{
-		con:        c,
-		remoteAddr: c.RemoteAddr(),
-		br:         br,
+	return &connDecoder{
+		Conn:       c,
+		ByteReader: br,
 		Decoder:    codec.NewDecoder(br, mh),
 	}
 }
 
-func (t *Transport) IsConnected() bool {
-	t.mutex.Lock()
-	ret := (t.cpkg != nil)
-	t.mutex.Unlock()
-	return ret
-}
-
-func (t *Transport) GetRemoteAddr() (ret net.Addr) {
-	if t.cpkg != nil {
-		ret = t.cpkg.GetRemoteAddr()
-	}
-	return
+type Transport struct {
+	sync.Locker
+	ByteEncoder
+	cdec           *connDecoder
+	dispatcher     dispatcher
+	packetizer     *packetizer
+	log            LogInterface
+	wrapError      WrapErrorFunc
+	writeCh        chan []byte
+	readByteCh     chan struct{}
+	decodeCh       chan interface{}
+	readerResultCh chan interface{}
+	writerResultCh chan error
+	startedCh      chan struct{}
+	stopCh         chan struct{}
 }
 
 func NewTransport(c net.Conn, l LogFactory, wef WrapErrorFunc) *Transport {
-	mh := codec.MsgpackHandle{WriteExt: true}
-
-	buf := new(bytes.Buffer)
-	ret := &Transport{
-		mh:        &mh,
-		cpkg:      NewconPackage(c, &mh),
-		buf:       buf,
-		enc:       codec.NewEncoder(buf, &mh),
-		mutex:     new(sync.Mutex),
-		rdlck:     new(sync.Mutex),
-		wrlck:     new(sync.Mutex),
-		wrapError: wef,
-	}
+	mh := &codec.MsgpackHandle{WriteExt: true}
+	cdec := newConnDecoder(c, mh)
 	if l == nil {
 		l = NewSimpleLogFactory(nil, nil)
 	}
-	log := l.NewLog(ret.cpkg.GetRemoteAddr())
-	ret.log = log
+	log := l.NewLog(cdec.RemoteAddr())
+	byteEncoder := NewFramedMsgpackEncoder(mh)
+
+	ret := &Transport{
+		Locker:         new(sync.Mutex),
+		ByteEncoder:    byteEncoder,
+		cdec:           cdec,
+		log:            log,
+		wrapError:      wef,
+		writeCh:        make(chan []byte),
+		readByteCh:     make(chan struct{}),
+		decodeCh:       make(chan interface{}),
+		readerResultCh: make(chan interface{}),
+		writerResultCh: make(chan error),
+		startedCh:      make(chan struct{}, 1),
+		stopCh:         make(chan struct{}),
+	}
 	ret.dispatcher = NewDispatch(ret, log, wef)
 	ret.packetizer = NewPacketizer(ret.dispatcher, ret)
+	// Make one token available to start
+	select {
```

Comment:
why select instead of just: `ret.startedCh <- struct{}{}`? Since `startedCh` was created with a buffer.